src/setup_payload/tests/TestAdditionalDataPayload.cpp

```diff
@@ -0,0 +1,197 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements a unit test suite for the additional payload generation
+ *      code functionality.
+ *
+ */
+
+#include <nlunit-test.h>
+#include <stdio.h>
+
+#include <setup_payload/AdditionalDataPayloadGenerator.cpp>
+#include <setup_payload/AdditionalDataPayloadParser.cpp>
+#include <setup_payload/SetupPayload.cpp>
+#include <setup_payload/SetupPayload.h>
+#include <system/SystemPacketBuffer.h>
+
+#include <lib/support/UnitTestRegistration.h>
+#include <lib/support/verhoeff/Verhoeff.h>
+
+using namespace chip;
+
+namespace {
+
+void TestGeneratingAdditionalDataPayloadWithRotatingId(nlTestSuite * inSuite, void * inContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferHandle bufferHandle;
+
+    char serialNumber[]      = "89051AAZZ236";
+    uint16_t lifetimeCounter = 10;
+    BitFlags<AdditionalDataFields> additionalDataFields;
+    additionalDataFields.Set(AdditionalDataFields::RotatingDeviceId);
+
+    // Init Chip memory management before the stack
+    chip::Platform::MemoryInit();
+
+    err = AdditionalDataPayloadGenerator().generateAdditionalDataPayload(lifetimeCounter, serialNumber, ArraySize(serialNumber),
+                                                                         bufferHandle, additionalDataFields);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !bufferHandle.IsNull());
+
+    char output[51];
+    err = BytesToUppercaseHexString(bufferHandle->Start(), bufferHandle->DataLength(), output, ArraySize(output));
+
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, strcmp(output, "5500003000120A00D1414E39C9A08469DA2E29F0ACAA086618") == 0);
+}
+
+void TestGeneratingAdditionalDataPayloadWithoutRotatingId(nlTestSuite * inSuite, void * inContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferHandle bufferHandle;
+
+    char serialNumber[]      = "89051AAZZ236";
+    uint16_t lifetimeCounter = 10;
+    BitFlags<AdditionalDataFields> additionalDataFields;
+
+    // Init Chip memory management before the stack
+    chip::Platform::MemoryInit();
+
+    err = AdditionalDataPayloadGenerator().generateAdditionalDataPayload(lifetimeCounter, serialNumber, ArraySize(serialNumber),
+                                                                         bufferHandle, additionalDataFields);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !bufferHandle.IsNull());
+
+    char output[51];
+    err = BytesToUppercaseHexString(bufferHandle->Start(), bufferHandle->DataLength(), output, ArraySize(output));
+
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, strcmp(output, "55000018") == 0);
+}
+
+void TestGeneratingAdditionalDataPayloadWithRotatingIdAndParsingIt(nlTestSuite * inSuite, void * inContext)
```

Comment:
This entirely repeats a previous test, which is a lot of boilerplate. Parsing SHOULD NOT be based on round-tripping, but rather on expected correct values, since otherwise we are only verifying end-to-end bug compatibility.

Reply:
refactored those test cases to minimum the repetitive code